File: collect_app/src/test/java/org/odk/collect/android/preferences/AdminPasswordDialogFragmentTest.java
package org.odk.collect.android.preferences;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import android.os.Bundle;

import androidx.fragment.app.FragmentActivity;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.odk.collect.android.R;
import org.odk.collect.android.injection.config.AppDependencyModule;
import org.odk.collect.android.support.RobolectricHelpers;
import org.odk.collect.android.support.TestActivityScenario;
import org.odk.collect.android.utilities.AdminPasswordProvider;
import org.robolectric.annotation.LooperMode;

import static android.os.Looper.getMainLooper;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.nullValue;
import static org.odk.collect.android.preferences.AdminPasswordDialogFragment.ARG_ACTION;
import static org.odk.collect.android.preferences.AdminPasswordDialogFragment.Action.ADMIN_SETTINGS;
import static org.robolectric.Shadows.shadowOf;
import static org.robolectric.annotation.LooperMode.Mode.PAUSED;

@RunWith(AndroidJUnit4.class)
@LooperMode(PAUSED)
public class AdminPasswordDialogFragmentTest {

    @Before
    public void setup() {
        RobolectricHelpers.overrideAppDependencyModule(new AppDependencyModule() {
            @Override
            public AdminPasswordProvider providesAdminPasswordProvider(PreferencesDataSourceProvider preferencesDataSourceProvider) {
                return new StubAdminPasswordProvider();
            }
        });
    }

    @Test
    public void enteringPassword_andClickingOK_callsOnCorrectAdminPasswordWithAction() {
        TestActivityScenario<SpyAdminPasswordDialogCallbackActivity> activityScenario = TestActivityScenario.launch(SpyAdminPasswordDialogCallbackActivity.class);
        activityScenario.onActivity(activity -> {
            AdminPasswordDialogFragment fragment = createFragment();
            fragment.show(activity.getSupportFragmentManager(), "tag");
            shadowOf(getMainLooper()).idle();

            fragment.getInput().setText("password");
            ((AlertDialog) fragment.getDialog()).getButton(AlertDialog.BUTTON_POSITIVE).performClick();
            shadowOf(getMainLooper()).idle();

            assertThat(activity.onCorrectAdminPasswordCalledWith, equalTo(ADMIN_SETTINGS));
            assertThat(activity.onIncorrectAdminPasswordCalled, equalTo(false));
        });
    }

    @Test
    public void enteringIncorrectPassword_andClickingOK_callsOnInCorrectAdminPassword() {
        TestActivityScenario<SpyAdminPasswordDialogCallbackActivity> activityScenario = TestActivityScenario.launch(SpyAdminPasswordDialogCallbackActivity.class);
        activityScenario.onActivity(activity -> {
            AdminPasswordDialogFragment fragment = createFragment();
            fragment.show(activity.getSupportFragmentManager(), "tag");
            shadowOf(getMainLooper()).idle();

            fragment.getInput().setText("not the password");
            ((AlertDialog) fragment.getDialog()).getButton(AlertDialog.BUTTON_POSITIVE).performClick();
            shadowOf(getMainLooper()).idle();

            assertThat(activity.onCorrectAdminPasswordCalledWith, nullValue());
            assertThat(activity.onIncorrectAdminPasswordCalled, equalTo(true));
        });
    }

    @Test
    public void afterRecreating_enteringPassword_andClickingOK_callsOnCorrectAdminPasswordWithAction() {
        TestActivityScenario<SpyAdminPasswordDialogCallbackActivity> activityScenario = TestActivityScenario.launch(SpyAdminPasswordDialogCallbackActivity.class);
        activityScenario.onActivity(activity -> {
            AdminPasswordDialogFragment fragment = createFragment();
            fragment.show(activity.getSupportFragmentManager(), "tag");
            shadowOf(getMainLooper()).idle();
        });

        activityScenario.recreate();
        activityScenario.onActivity(activity -> {
            AdminPasswordDialogFragment fragment = (AdminPasswordDialogFragment) activity.getSupportFragmentManager().findFragmentByTag("tag");

            fragment.getInput().setText("password");
            ((AlertDialog) fragment.getDialog()).getButton(AlertDialog.BUTTON_POSITIVE).performClick();
            shadowOf(getMainLooper()).idle();

            assertThat(activity.onCorrectAdminPasswordCalledWith, equalTo(ADMIN_SETTINGS));
            assertThat(activity.onIncorrectAdminPasswordCalled, equalTo(false));
        });
    }

    private AdminPasswordDialogFragment createFragment() {
        AdminPasswordDialogFragment fragment = new AdminPasswordDialogFragment();
        Bundle args = new Bundle();
        args.putSerializable(ARG_ACTION, ADMIN_SETTINGS);
        fragment.setArguments(args);
        return fragment;
    }

    private static class StubAdminPasswordProvider extends AdminPasswordProvider {

        StubAdminPasswordProvider() {
            super(null);
        }

        @Override
        public boolean isAdminPasswordSet() {
            return true;
        }

        @Override
        public String getAdminPassword() {
            return "password";
        }
    }

    private static class SpyAdminPasswordDialogCallbackActivity extends FragmentActivity implements AdminPasswordDialogFragment.AdminPasswordDialogCallback  {

        private AdminPasswordDialogFragment.Action onCorrectAdminPasswordCalledWith;
        private boolean onIncorrectAdminPasswordCalled;

        @Override
        protected void onCreate(@Nullable Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            setTheme(R.style.Theme_AppCompat); // Needed for androidx.appcompat.app.AlertDialog
        }

        @Override
        public void onCorrectAdminPassword(AdminPasswordDialogFragment.Action action) {
            onCorrectAdminPasswordCalledWith = action;
        }

        @Override
        public void onIncorrectAdminPassword() {
            onIncorrectAdminPasswordCalled = true;
        }
    }
}

File: collect_app/src/test/java/org/odk/collect/android/preferences/dialogs/AdminPasswordDialogFragmentTest.kt
package org.odk.collect.android.preferences.dialogs

import android.text.InputType
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.ViewModel
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.pressBack
import androidx.test.espresso.matcher.ViewMatchers.isRoot
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.`is`
import org.hamcrest.Matchers.nullValue
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoMoreInteractions
import org.mockito.kotlin.whenever
import org.odk.collect.android.injection.config.AppDependencyModule
import org.odk.collect.android.preferences.ProjectPreferencesViewModel
import org.odk.collect.android.support.CollectHelpers
import org.odk.collect.android.utilities.AdminPasswordProvider
import org.odk.collect.fragmentstest.FragmentScenarioLauncherRule
import org.odk.collect.settings.SettingsProvider
import org.odk.collect.testshared.RobolectricHelpers
import javax.inject.Inject

@RunWith(AndroidJUnit4::class)
class AdminPasswordDialogFragmentTest {

    private val adminPasswordProvider = mock<AdminPasswordProvider>()
    private val projectPreferencesViewModel = mock<ProjectPreferencesViewModel>()

    @Inject
    lateinit var factory: ProjectPreferencesViewModel.Factory

    @get:Rule
    val launcherRule = FragmentScenarioLauncherRule()

    @Before
    fun setup() {
        CollectHelpers.overrideAppDependencyModule(object : AppDependencyModule() {
            override fun providesAdminPasswordProvider(settingsProvider: SettingsProvider): AdminPasswordProvider {
                return adminPasswordProvider
            }

            override fun providesProjectPreferencesViewModel(adminPasswordProvider: AdminPasswordProvider): ProjectPreferencesViewModel.Factory {
                return object : ProjectPreferencesViewModel.Factory(adminPasswordProvider) {
                    override fun <T : ViewModel> create(modelClass: Class<T>): T {
                        return projectPreferencesViewModel as T
                    }
                }
            }
        })
    }

    @Test
    fun `The dialog should be dismissed after clicking on a device back button`() {
        val scenario = launcherRule.launchDialogFragment(AdminPasswordDialogFragment::class.java)
        scenario.onFragment {
            assertThat(it.dialog!!.isShowing, `is`(true))
            onView(isRoot()).perform(pressBack())
            assertThat(it.dialog, `is`(nullValue()))
        }
    }

    @Test
    fun `The dialog should be dismissed after clicking on 'OK'`() {
        val scenario = launcherRule.launchDialogFragment(AdminPasswordDialogFragment::class.java)
        scenario.onFragment {
            assertThat(it.dialog!!.isShowing, `is`(true))
            (it.dialog as AlertDialog?)!!.getButton(AlertDialog.BUTTON_POSITIVE).performClick()
            RobolectricHelpers.runLooper()
            assertThat(it.dialog, `is`(nullValue()))
        }
    }

    @Test
    fun `The dialog should be dismissed after clicking on 'CANCEL'`() {
        val scenario = launcherRule.launchDialogFragment(AdminPasswordDialogFragment::class.java)
        scenario.onFragment {
            assertThat(it.dialog!!.isShowing, `is`(true))
            (it.dialog as AlertDialog?)!!.getButton(AlertDialog.BUTTON_NEGATIVE).performClick()
            RobolectricHelpers.runLooper()
            assertThat(it.dialog, `is`(nullValue()))
        }
    }

    @Test
    fun `Entering correct password sets Unlocked state in view model`() {
        whenever(adminPasswordProvider.isAdminPasswordSet).thenReturn(true)
        whenever(adminPasswordProvider.adminPassword).thenReturn("password")

        val scenario = launcherRule.launchDialogFragment(AdminPasswordDialogFragment::class.java)
        scenario.onFragment {
            it.binding.editText.setText("password")
            (it.dialog as AlertDialog?)!!.getButton(AlertDialog.BUTTON_POSITIVE).performClick()
            RobolectricHelpers.runLooper()
            verify(projectPreferencesViewModel).setStateUnlocked()
            verifyNoMoreInteractions(projectPreferencesViewModel)
        }
    }

    @Test
    fun `Entering incorrect password does not set any state in view model`() {
        whenever(adminPasswordProvider.isAdminPasswordSet).thenReturn(true)
        whenever(adminPasswordProvider.adminPassword).thenReturn("password")

        val scenario = launcherRule.launchDialogFragment(AdminPasswordDialogFragment::class.java)
        scenario.onFragment {
            it.binding.editText.setText("incorrect_password")
            (it.dialog as AlertDialog?)!!.getButton(AlertDialog.BUTTON_POSITIVE).performClick()
            RobolectricHelpers.runLooper()
            verifyNoMoreInteractions(projectPreferencesViewModel)
        }
    }

    @Test
    fun `When screen is rotated password and checkbox value is retained`() {
        val scenario = launcherRule.launchDialogFragment(AdminPasswordDialogFragment::class.java)
        scenario.onFragment {
            it.binding.editText.setText("password")
            it.binding.checkBox.performClick()
            scenario.recreate()
            assertThat(it.binding.editText.text.toString(), `is`("password"))
            assertThat(it.binding.checkBox.isChecked, `is`(true))
        }
    }

    @Test
    fun `'Show password' displays and hides password`() {
        val scenario = launcherRule.launchDialogFragment(AdminPasswordDialogFragment::class.java)
        scenario.onFragment {
            it.binding.checkBox.performClick()
            assertThat(it.binding.editText.inputType, `is`(InputType.TYPE_TEXT_VARIATION_PASSWORD))
            it.binding.checkBox.performClick()
            assertThat(it.binding.editText.inputType, `is`(InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD))
        }
    }
}
